### PR TITLE
Skip Constant Folding for ops producing an optional type output

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: Optional tensors that contain data are tensors at runtime and hence the check that an output OrtValue of an op be a tensor before deciding whether to constant fold the op is not correct. Instead, make the decision based on the proto type of the output NodeArg and ensure that the type is of type tensor before going ahead with constant folding. 

**Motivation and Context**
Fixes https://github.com/microsoft/onnxruntime/issues/11717

### CODE REVIEW

File: onnxruntime/core/optimizer/constant_folding.cc [hunk 1]
```diff
@@ -196,10 +196,10 @@ Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level,
       ORT_ENFORCE(fetches.size() == node->OutputDefs().size());
       converted_to_constant = true;
       for (size_t fetch_idx = 0; fetch_idx < fetches.size(); ++fetch_idx) {
-        OrtValue& ort_value = fetches[fetch_idx];
+        auto* constant_arg_out = node->MutableOutputDefs()[fetch_idx];
         // XXX: Add support for SparseTensors outputs when we have sparse outputs
-        if (!ort_value.IsTensor()) {
-          LOGS(logger, WARNING) << "Unsupported output type of " << ort_value.Type()
+        if (!utils::HasTensorType(*constant_arg_out->TypeAsProto())) {
+          LOGS(logger, INFO) << "Unsupported output type of " << constant_arg_out->Type()
```
Comment (hariharans29) [inline onnxruntime/core/optimizer/constant_folding.cc:202]: Changed WARNING to INFO in order to make this less jarring as it came up in another issue

File: onnxruntime/core/providers/cpu/optional/optional_ops.cc [hunk 2]
```diff
@@ -118,7 +118,11 @@ Status Optional::Compute(OpKernelContext* ctx) const {
 
   } else {  // No input was provided - we use the type proto to construct the output OrtValue
 
-    CheckValidTypeProto(*type_proto_);
+    if (!CheckValidTypeProto(*type_proto_)) {
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/optional/optional_ops.cc:121]: Minor misc change that was missing in the original check-in

File: onnxruntime/core/providers/cpu/optional/optional_ops.cc [hunk 3]
```diff
@@ -118,7 +118,11 @@ Status Optional::Compute(OpKernelContext* ctx) const {
...
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "The TypeProto attribute in the Optional op ",
+                             "can only be of type(tensor) or (seq(tensor))");
+    }
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/optional/optional_ops.cc:126]: Does the ONNX validation (either type/shape inferencing or the type constraints for the operator) not check this?

File: onnxruntime/core/optimizer/constant_folding.cc [hunk 4]
```diff
@@ -196,10 +196,10 @@ Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level,
       ORT_ENFORCE(fetches.size() == node->OutputDefs().size());
       converted_to_constant = true;
       for (size_t fetch_idx = 0; fetch_idx < fetches.size(); ++fetch_idx) {
-        OrtValue& ort_value = fetches[fetch_idx];
+        auto* constant_arg_out = node->MutableOutputDefs()[fetch_idx];
         // XXX: Add support for SparseTensors outputs when we have sparse outputs
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/constant_folding.cc:200]: do we need a mutable NodeArg instance here?

nit: would be slightly nicer to use NodeArg& to make it clear you don't expect a nullptr to be returned by the call to get the NodeArg.

File: onnxruntime/core/providers/cpu/optional/optional_ops.cc (hunk 3 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/optional/optional_ops.cc:126]: The type constraints for the operator can't check this as this is an attribute (which is a generic TypeProto). The type/shape inferencing seems to just copy over the provided tp as the output tp when it probably should have also checked to ensure that the provided tp is a tensor or a seq(tensor) given that the output type of the op can only be those types - https://github.com/onnx/onnx/blob/805ae1e634697e37b43701e585c9c253a29ce076/onnx/defs/optional/defs.cc#L53.

File: onnxruntime/core/optimizer/constant_folding.cc (hunk 4 above)
Comment (hariharans29) [reply onnxruntime/core/optimizer/constant_folding.cc:200]: Done

**Comment by skottmckay** [comment]
Is the intention to add the unit test before checkin?

**Comment by hariharans29** [comment]
> Is the intention to add the unit test before checkin?

Yes. Waiting for the OP to run tests against this branch. I will add a test once it is determined that this is all there is to fix at this point.

**Comment by hariharans29** [comment]
> > Is the intention to add the unit test before checkin?
> 
> Yes. Waiting for the OP to run tests against this branch. I will add a test once it is determined that this is all there is to fix at this point.

@skottmckay - I have added a test for this, can you please take a look ?

File: onnxruntime/test/shared_lib/test_inference.cc [hunk 5]
```diff
@@ -2164,3 +2165,13 @@ TEST(CApiTest, TestCudaMemcpyToHostWithSequenceTensors) {
...
 #endif
+
+#if !defined(DISABLE_OPTIONAL_TYPE)
+TEST(CApiTest, GH_11717) {
+  const auto* model_path = OPTIONAL_TYPE_GH_11717_MODEL;
+  Ort::SessionOptions session_options{};
+  // Just check if the model loads fine without a segmentation fault
+  // in the default CPU EP
+  Ort::Session session{*ort_env, model_path, session_options};
+}
```
Comment (skottmckay) [inline onnxruntime/test/shared_lib/test_inference.cc:2176]: EXPECT_NO_THROW?

**Review by skottmckay — APPROVED** [review]
:shipit:

File: onnxruntime/test/shared_lib/test_inference.cc (hunk 5 above)
Comment (hariharans29) [reply onnxruntime/test/shared_lib/test_inference.cc:2176]: Done

**Review by skottmckay — APPROVED** [review]

**Pull request merged**